### DISCUSSION TITLE
File, note, link UI prefix not displaying

### DISCUSSION THREAD

**Discussion by claudejrogers** [emacs-citar/citar#621] [Q&A] [answered]
Hi, the UI "prefix" indicating which entries have files, notes, or links is not displaying for me (emacs 28.1). When I run `citar-open` and select an entry with a pdf, I get a menu to select opening the link or pdf, and I've configured the library and note paths variables. Any suggestions?

**Comment by bdarcus** [comment]
What note extension are you using?

And the file names correspond to the citekeys, something like the below?

![image](https://user-images.githubusercontent.com/1134/171403475-68ebe857-2654-4764-8a66-7ca1c4881a87.png)

**Reply by claudejrogers** [reply]
I am using org-roam-bibtex, but I don't have many notes yet.
The issue for me is the pdfs. I don't have any indicators for which entries have pdfs using citar, and there isn't even space for the "prefix" in the UI:
<img width="1792" alt="citar" src="https://user-images.githubusercontent.com/614117/171422261-3c507c5d-71c1-435d-bd95-f0dce9404728.png">

But it's working in ivy-bibtex:
<img width="1792" alt="ivy-bibtex" src="https://user-images.githubusercontent.com/614117/171422368-39d24de0-05ef-4a54-a919-376882ddbcff.png">

And the pdf filenames are the citekeys:
<img width="568" alt="pdfs" src="https://user-images.githubusercontent.com/614117/171422486-959793ad-7aae-46ae-afcb-03a03f33e85c.png">

**Reply by bdarcus** [reply]
> ... there isn't even space for the "prefix" in the UI

Hmm ... that's really weird. It means the affixation function isn't working for some reason. I've never seen that, and can't imagine why.

If that display function was working, but citar was just not correctly associating the files, you would have an empty space there.

You're using vertico?

**Reply by claudejrogers** [reply]
Yes, vertico is enabled (I'm using doom emacs) and ripgrep is installed.

**Reply by bdarcus** [reply]
So same as me (though I am on Emacs 29).

Is it possible you have a buggy emacs install; say built from source from the git repo or something?

Can you check the messages to see if Emacs reports anything related to that?

**Reply by claudejrogers** [reply]
I've checked that there's nothing in messages.

**Reply by claudejrogers** [reply]
It could be an issue with ripgrep, here is some of the `doom doctor` output:
```
 > :completion vertico
          ! The installed ripgrep binary was not built with support for PCRE lookaheads.
            Some advanced consult filtering features will not work as a result, see the module readme.
          ! The installed grep binary was not built with support for PCRE lookaheads.
            Some advanced consult filtering features will not work as a result, see the module readme.
```

**Reply by bdarcus** [reply]
I think we need to do some basic debugging; for example, to see if other software that uses affixation-functions works correctly, or not.

Doom uses the `all-the-icons-completion` package, which does a similar thing with icons. Do you by chance have that `(+icons)` turned on? If not, maybe try it and see what happens?

Or if it IS on, turn it off, since in the past I recall a conflict with that package that I thought its author fixed.

That would hopefully allow us to narrow down what the problem is. I'm pretty sure it's not citar though.

EDIT: I doubt ripgrep is the issue.

**Reply by claudejrogers** [reply]
I turned off the icons ran `doom purge` then `doom sync` and it didn't change citar output. I re-enabled +icons re-synced and it didn't help.
Is there any way to check the affixation-function output in an elisp buffer? Or maybe check if any entries evaluate as true for has-file?

**Reply by claudejrogers** [reply]
Ok, running citar-has-file in eshell has sensible output

**Reply by bdarcus** [reply]
> Is there any way to check the affixation-function output in an elisp buffer? 

```elisp
 (citar--affixation (mapcar (lambda (c) (car c)) (citar--get-candidates)))
```

Here's example of the prefix output (with icons):

```elisp
   #("          " 0 1
     (face
      (:family "FontAwesome" :height 1.2 :inherit all-the-icons-green)
      font-lock-face
      (:family "FontAwesome" :height 1.2 :inherit all-the-icons-green)
      display
      (raise -0.12)
      rear-nonsticky t)
     2 3
     (display
      (space :align-to 3))
     5 6
     (display
      (space :align-to 6))
     8 9
     (display
      (space :align-to 9))
     10 11
     (display
      (space :align-to 11)))
   "")
```

You can also do (edited a mistake below):

```elisp
(member "citekey" (citar-file--keys-with-library-files))
```

**Reply by claudejrogers** [reply]
Running `(citar--affixation (mapcar (lambda (c) (car c)) (citar--get-candidates)))`
produced expected output (but the file icon is rendering weirdly):
<img width="1640" alt="Screen Shot 2022-06-01 at 9 12 22 AM" src="https://user-images.githubusercontent.com/614117/171451765-0b5107bf-9707-480e-a496-c605fb9f2617.png">


Running `(member "citekey" (citar-file--keys-with-library-files))`
gave `member: Symbol’s function definition is void: citar-file--keys-with-library-files`

**Reply by bdarcus** [reply]
That's actually expected.

The second example is new, and the Doom module probably hasn't been bumped yet.

But that confirms the affixation function works correctly. 

I'll need to think about this.

**Reply by claudejrogers** [reply]
Thanks for all your help!

**Reply by bdarcus** [reply]
Wait, though; is that output actually correct? Do you see a has-file indicator in the right place?

**Reply by claudejrogers** [reply]
Good catch. The indicators seem to be appended as suffixes with lots of whitespace.

**Reply by claudejrogers** [reply]
It's hard to say what exactly is going on because I ran it in org mode and it does that table formatting.

**Reply by bdarcus** [reply]
Maybe try in `ielm` just to be sure?

**Reply by claudejrogers** [reply]
I may have found the issue?

Here is the list output:
<img width="1792" alt="Screen Shot 2022-06-01 at 9 36 58 AM" src="https://user-images.githubusercontent.com/614117/171455883-b1e799bc-0935-4fe5-ad7c-b7e9fdad226b.png">
The first entry has a link and no file, the second entry has both a link and a file, which is correct. But `invisible` is set to true.

**Reply by bdarcus** [reply]
The text `has:files` is deliberately set to invisible. That enables you to type, for example, `has:files` or even `:f` in the prompt, and for it to filter based on that.

The affixation is just for the visual display.

But from what I can see, the output is right.

I'm out of ideas at this point.

To go back to above:

> I turned off the icons ran doom purge then doom sync and it didn't change citar output. 

And you of course rebooted, which would force recreation of the candidates?

> I re-enabled +icons re-synced and it didn't help.

But do you see the `all-the-icons-completions` affixation icons, outside of citar?

E.g. is affixation broken everywhere for you, or only in citar?

**Reply by claudejrogers** [reply]
What else runs affixation? If I run IBuffer or dired with +icons I see icons in the output.

**Reply by bdarcus** [reply]
On the left side, same as where the citar ones should be?

If as I assume yes, then I'm totally confused, because that suggests:

1. no bug in Emacs generally
2. no conflict with `all-the-icons-completion` (since you said turning it off didn't fix things in citar)
3. really haven't seen this in citar at all previously; I'm almost 100% confident there's no bug here

So I don't know what else it could possibly be.

@iyefrat - have you ever seen affixation just not displaying like this? Any ideas?

**Reply by claudejrogers** [reply]
The dired and IBuffer _buffers_ have icons in the correct location, but there are no icons in the completion buffer after running C-x C-f. In a language mode with lsp support, there are icons to the left of suggestions.

**Reply by claudejrogers** [reply]
No icons running SPC b b either

**Reply by bdarcus** [reply]
`find-file`; a la the screenshot here?

https://github.com/iyefrat/all-the-icons-completion

**Reply by claudejrogers** [reply]
Ah, I found the issue. I had ivy enabled in the completion section in init.el, in addition to vertico +icons. Removing ivy, purging, and syncing fixed citar and find-file, SPC b b, etc. Thanks for the help.

**Reply by bdarcus** [reply]
Ugh, I thought to ask that earlier, but forgot!

**Reply by claudejrogers** [reply]
Yeah, I'm embarrassed I didn't notice earlier.

**Comment by bdarcus** [comment, marked as answer]
If you turn on the "vertico" module in doom, then turn off "ivy".
